src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -45,5 +50,289 @@ protected override void InsertItem(int index, IDeltaSetItem item)
+        {
+            //Here we are getting the keys and using the keys to find the original object 
+            //to patch from the list of collection
+
+            ODataAPIHandler<TStructuralType> apiHandlerOfT = apiHandler as ODataAPIHandler<TStructuralType>;
+
+            Debug.Assert(apiHandlerOfT != null);
+
+            DeltaSet<TStructuralType> deltaSet = CreateDeltaSet();
+
+            foreach (Delta<TStructuralType> changedObj in Items)
+            {
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+
+                //Get filtered item based on keys
+                TStructuralType original = null;
+                string errorMessage = string.Empty;
+                string getErrorMessage = string.Empty;
+
+                Dictionary<string, object> keyValues = new Dictionary<string, object>();
+
+                foreach (string key in _keys)
+                {
+                    object value;
+
+                    if (changedObj.TryGetPropertyValue(key, out value))                   
+                    {
+                        keyValues.Add(key, value);
+                    }
+                }
+
+                try
+                {
+                    ODataAPIResponseStatus ODataAPIResponseStatus = apiHandlerOfT.TryGet(changedObj.ODataPath.GetKeys(), out original, out getErrorMessage);
+
+                    DeltaDeletedEntityObject<TStructuralType> deletedObj = changedObj as DeltaDeletedEntityObject<TStructuralType>;
+
+                    if (ODataAPIResponseStatus == ODataAPIResponseStatus.Failure || (deletedObj != null && ODataAPIResponseStatus == ODataAPIResponseStatus.NotFound))
+                    {
+                        IDeltaSetItem deltaSetItem = changedObj;
+
+                        DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+                        dataModificationExceptionType.MessageType = new MessageType { Message = getErrorMessage };
+
+                        deltaSetItem.TransientInstanceAnnotationContainer.AddResourceAnnotation(SRResources.DataModificationException, dataModificationExceptionType);
+
+                        deltaSet.Add(deltaSetItem);
+                        
+                        continue;
+                    }
+
+                    if (deletedObj != null)
+                    {
+                        operation = DataModificationOperationKind.Delete;
+
+                        changedObj.CopyChangedValues(original, apiHandlerOfT, apiHandlerFactory);
+
+                        if (apiHandlerOfT.TryDelete(keyValues, out errorMessage) != ODataAPIResponseStatus.Success)
+                        {
+                            //Handle Failed Operation - Delete                           
+                            
+                            if (ODataAPIResponseStatus == ODataAPIResponseStatus.Success)
```

Comment:
I think this check is superfluous. From the `if ..else` above, the `ODataAPIResponseStatus` cannot be a `success`

Reply:
The `ODataAPIResponseStatus` is from the `TryGet` above.
Before this section, we have handled scenario where we have a delete and the object is not found using `TryGet`.
So we are here because the response status from TryGet is Success. This code only fool proofs the logic just incase we add another response status in the future (which looks unlikely). For now this check is not needed but I can keep it. What do you think?